### ISSUE TITLE
Initial Spec

### ISSUE THREAD

**Issue by neetjn** [neetjn/py-component-controller#2]
`py-component-controller` is a fork from my PMC (page modal architecture) working at Transparent Language. It should radically simplify writing and maintaining selenium tests with python.

`py-component-controller` is going to provide a wdio like interface in regards to searching for WebElements. Rather than raising a `NoSuchElementException` when elements aren't found, component properties will simply return `None`. This makes it much easier to check for element availability, rather than having to wrap over exception cases.

The `@element` decorator should detect the selector and appropriately invoke `find_element_by_css_selector` or `find_element_by_xpath`. Additionally, selectors will support formating templates.

An example would be,

```python
from pyscc import Component

class Profile(Component):

  @element
  def save_button(self):
    return 'button[ng-click="onSave()"]'

  @element
  def tooltip(self):
    return 'span.tooltip[aria-label="{label}"]'

profile = Profile(webdriver)
profile.save_button.get() >> return WebElement or None
# driver.find_element_by_css_selector('selector')
profile.tooltip.get.fmt(label="Hello World") >> return WebElement or None
# driver.find_element_by_css_selector('selector'.format(**kwargs))
profile.save_button.is.available() >> return True or False
# try:
#   driver.find_element_by_css_selector('selector')
# except WebElementException:
#   return False
# else:
#    return True
profile.save_button.is.visible() >> return True or False
# try:
#   controller.javascript.is_visible(driver.find_element_by_css_selector('selector'))
# except WebElementException:
#   return False
# else:
#    return True
profile.save_button.scroll_to() >> return True or False
...
```

These apis should also simplify existing `Controller` functionality, consuming callable methods from component elements.

Legacy,
```python
controller.wait(timeout=5, condition=lambda: controller.element_exists(expression=lambda: page.button))
```

Pyscc,
```python
controller.wait(timeout=5, component.element.is.visible)
```

**Comment by neetjn** [comment]
The `@element` decorator is just going to hijack the function, get the selector and determine whether to use xpath or css. It will then return some `Element` object that will contain ...

To pull this all off, the `Controller` can pass a reference to itself to each `Component`, so instances of each component have access to the controller's functionality and javascript library. We could also just scrap away our currency logic and do away with passing in a logger, env, and driver reference.

**Comment by neetjn** [comment]
Here's a light demo,

```python
import inspect
from six import iteritems

# definitions ======================================================

class Resource(object):
  def __init__(self, **kwargs):
    for prop, val in iteritems(kwargs):
      setattr(self, prop, val)

class Controller(object):
  def __init__(self, driver, components):
    self.driver = driver
    self.components = Resource(**{
      name: component(controller=self) for name, component in iteritems(components)})

class Component(object):
  def __init__(self, controller):
    self.controller = controller
    
class Element(object):
  def __init__(self, controller, selector):
    self.controller = controller
    self.selector = selector
    
  @property
  def get(self):
    return self.selector
    
  @property
  def is(self):
    return self.selector
    
def element(func):
  @property
  def wrapper(self):
    return Element(self.controller, func(self))
  return wrapper

# application ======================================================

class HomeComponent(Component):
  
  @element
  def button(self):
    return 'button[ng-click="test"]'
    
class AppController(Controller):
  
  def __init__(self):
    super(AppController, self).__init__('webdriver', {
      'home': HomeComponent
    })
  
  def home_button_selector(self):
    return self.components.home.button().get()

# demo ==============================================================
    
app = AppController()
print(app.components.home.button.get)
```

**Comment by neetjn** [comment]
We'll also need to add support properties for multiple elements, in which case we can probably introduce another decorator `@elements`.

**Comment by neetjn** [comment]
The `Element` wrapper should contain an api to wait for the element to be both available and visible.

Example,
```python
home = controller.components.home
username = home.username.wait_for(5)
password = home.password.wait_visible(5)
```

These waits should also optionally allow for an error to be raised.

```python
home.username.wait_for(timeout=5, error='Username field was not available')
```

**Comment by neetjn** [comment]
Additionally, the `Element` wrapper should include shortcut javascript utilities such as `get_attribute`, `get_property`, `trigger_event`, etc.

**Comment by neetjn** [comment]
The `Element` wrapper should also include an asynchronous javascript wait as done in pyselenium-js.

Proposed syntax,
```python
component.element\
  .wait_js('$el.getAttribute("aria-data") == "viewed"', 250)\
  .click()\
  .wait(timeout=10, condition=component.element.check.wait_status)
```

**Comment by neetjn** [comment]
For more complex elements "groups", a new decorator `@component_group` should be created. This decorator will take dictionaries of key name pairs for elements and selectors, and instantiate a `Resource` instance with scoped element wrappers.

**Comment by neetjn** [comment]
The `Element` wrapper should include a method `switch_to`, to switch into the context of a given iframe.

**Comment by neetjn** [comment]
The `Element` wrapper should include a method `get_text`, to pull text from a given element. This method should also take an optional parameter, `raw`, to extract the inner html of an element.

**Comment by neetjn** [comment]
Closing, all the features discussed in this issue has been implemented and version 0.0.1 has been released.
